### PR TITLE
refactor: edit models

### PR DESCRIPTION
**Pull request by gagarinkomar**
Changed lazy='dynamic' to DynamicMapped, it is the same thing but more convenient.
In Grade instead of User relationship using UserOnCourse.
Swapped TaskGroup and Task for convenience(link to course now in TaskGroup instead of Task).
Removed unused fields: is_manytask_admin, is_course_admin.
Add gitlab_instance_host field to course and validation gitlab_instance_host equal when creating UserOnCourse.
Also edited test_models.py

Current scheme:
https://drawsql.app/teams/gagarinkomar/diagrams/copy-of-copy-of-manytask

### CODE REVIEW

File: manytask/models.py [hunk 1]
```diff
@@ -62,13 +62,36 @@ class UserOnCourse(Base):
...
+            user = target.user
+        else:
+            user = session.query(User).filter_by(id=target.user_id).one()
+
+        if target.course:
+            course = target.course
+        else:
+            course = session.query(Course).filter_by(id=target.course_id).one()
+    except Exception:
+        return
```
Comment (feldsherov) [inline manytask/models.py:84]: Why do you need to swallow exception here?

You are catching exception and after that just return, this is antipattern, do we have a good reason for doing this?

File: manytask/models.py [hunk 2]
```diff
@@ -62,13 +62,36 @@ class UserOnCourse(Base):
...
+
+@event.listens_for(UserOnCourse, 'before_insert')
+def validate_gitlab_instance_host(mapper: Mapper[UserOnCourse], connection: Connection, target: UserOnCourse) -> None:
+    session = Session(bind=connection)
+
+    try:
+        if target.user:
+            user = target.user
+        else:
+            user = session.query(User).filter_by(id=target.user_id).one()
```
Comment (feldsherov) [inline manytask/models.py:77]: When do we expect to have missing user in target?
Drop a comment, why can it happen?

File: manytask/models.py [hunk 3]
```diff
@@ -62,13 +62,36 @@ class UserOnCourse(Base):
...
+
+        if target.course:
```
Comment (feldsherov) [inline manytask/models.py:79]: Same question about missing course.

**Review by feldsherov — CHANGES_REQUESTED** [review]

File: manytask/models.py [hunk 4]
```diff
@@ -26,23 +26,24 @@ class User(Base):
...
+    gitlab_instance_host: Mapped[str]
     registration_secret: Mapped[str]
     show_allscores: Mapped[bool] = mapped_column(default=False)
 
     # relationships
-    tasks: Mapped[List['Task']] = relationship(back_populates='course', lazy='dynamic')
-    users_on_courses: Mapped[List['UserOnCourse']] = relationship(
-        back_populates='course', lazy='dynamic')
+    task_groups: DynamicMapped['TaskGroup'] = relationship(
+        back_populates='course')
```
Comment (prawwtocol) [inline manytask/models.py:44]: don't we need to set up cascade options to properly handle orphaned records?
something like
`cascade="all, delete-orphan"`

File: manytask/models.py [hunk 5]
```diff
@@ -26,23 +26,24 @@ class User(Base):
...
+    users_on_courses: DynamicMapped['UserOnCourse'] = relationship(
+        back_populates='course')
```
Comment (prawwtocol) [inline manytask/models.py:46]: same question here

File: manytask/models.py [hunk 6]
```diff
@@ -79,40 +102,41 @@ class TaskGroup(Base):
...
     __tablename__ = 'grades'
 
     id: Mapped[int] = mapped_column(primary_key=True)
-    user_id: Mapped[int] = mapped_column(ForeignKey(User.id))
+    user_on_course_id: Mapped[int] = mapped_column(ForeignKey(UserOnCourse.id))
     task_id: Mapped[int] = mapped_column(ForeignKey(Task.id))
-    score: Mapped[int]
-    submit_date: Mapped[datetime]
+    score: Mapped[int] = mapped_column(default=0)
+    last_submit_date: Mapped[datetime] = mapped_column(default=lambda: datetime.now(timezone.utc))
```
Comment (prawwtocol) [inline manytask/models.py:134]: why not `func.now()`?

File: tests/test_models.py [hunk 7]
```diff
@@ -154,117 +172,154 @@ def test_task_group(session):
...
+    session.add_all([user, course, user_on_course, task_group, task])
     session.commit()
 
     grade = Grade(
-        user=user,
+        user_on_course=user_on_course,
         task=task,
         score=77,
-        submit_date=datetime.now(timezone.utc)
+        last_submit_date=datetime.now(timezone.utc)
```
Comment (prawwtocol) [inline tests/test_models.py:208]: to ensure consistency, create a fixture, for e.g. `current_time` and specify a fixed datetime there

File: tests/test_models.py [hunk 8]
```diff
@@ -133,9 +143,14 @@ def test_user_on_course_unique_ids(session):
 
 
 def test_deadline(session):
+    course = Course(name="course0001", registration_secret="secret",
+                    gitlab_instance_host='gitlab.inst.org')
+    session.add(course)
+    session.commit()
+
     deadline = Deadline(data={"test_key": "test_value"})
```
Comment (prawwtocol) [inline tests/test_models.py:151]: I think we should also test invalid json and empty data

File: manytask/models.py (hunk 1 above)
Comment (gagarinkomar) [reply manytask/models.py:84]: Because if there is a problem with finding user or course, we don't want to think about exceptions here. We swallow the exception so that sqlalchemy.orm raises the necessary exception for us when trying to create an object itself.

File: manytask/models.py (hunk 2 above)
Comment (gagarinkomar) [reply manytask/models.py:77]:
```
user_on_course = UserOnCourse(
    user_id=1,
    course=course,
    repo_name="user_repo"
)
```

If we create an object like this, we have user_id instead of user

File: manytask/models.py (hunk 3 above)
Comment (gagarinkomar) [reply manytask/models.py:79]: Similar example

File: manytask/models.py (hunk 4 above)
Comment (gagarinkomar) [reply manytask/models.py]: Added cascade options and tests for them

File: manytask/models.py (hunk 6 above)
Comment (gagarinkomar) [reply manytask/models.py]: Changed this

File: tests/test_models.py (hunk 7 above)
Comment (gagarinkomar) [reply tests/test_models.py]: Created fixture and refactor code parts with current time usage

File: tests/test_models.py (hunk 8 above)
Comment (gagarinkomar) [reply tests/test_models.py:156]: Add test_deadline_date for test it

File: manytask/models.py (hunk 2 above)
Comment (prawwtocol) [reply manytask/models.py:85]: @gagarinkomar yeah I think the point was that you were advised to include that exact comment in code -- so that anyone reading this method would understand why target.user could be None

File: manytask/models.py (hunk 1 above)
Comment (prawwtocol) [reply manytask/models.py:92]: @gagarinkomar I think the worst part of this is catching a The Most General Exception Posssible (TM). Maybe NoResultFound would be better?

**Review by akostrikov — APPROVED** [review]
LGTM

File: manytask/models.py (hunk 1 above)
Comment (akostrikov) [reply manytask/models.py:92]: @gagarinkomar can you add logging here and a todo?
Seems like a funny place to swallow errors later.

Comment (akostrikov) [reply manytask/models.py:92]: https://github.com/manytask/manytask/issues/219 added issue
